Scripts/prepare_release.py

```diff
@@ -57,24 +56,52 @@ def unwrap(opt: Optional[A]) -> A:
     assert opt
     return opt
 
+class NewsFragment(NamedTuple):
+    pr: str
+    contents: str
+    path: Optional[Path]
+
+    @classmethod
+    def from_file(cls, pth: Path) -> "NewsFragment":
+        return NewsFragment(pth.stem, pth.read_text(encoding="utf-8"), pth)
+
 class NewsFragments:
+    r"""A collection of release note entries.
+
+    >>> import tempfile
+    >>> with tempfile.TemporaryDirectory() as tmpdir:
+    ...   fpath = Path(tmpdir) / "1234.fix"
+    ...   _ = fpath.write_text("Improved toaster settings.\nDafny will not burn toast again.", encoding="utf-8")
+    ...   print(NewsFragments.from_directory(tmpdir).render())
+    ## Bug fixes
+    <BLANKLINE>
+    - Improved toaster settings.
+      Dafny will not burn toast again.
+      (https://github.com/dafny-lang/dafny/pull/1234)
```

Comment:
Thanks for the testing code ! It helps a lot.

```suggestion
    - Improved toaster settings.
      Dafny will not burn toast again (https://github.com/dafny-lang/dafny/pull/1234)
```
```suggestion
    - Improved toaster settings.
      Dafny will not burn toast again.
      (https://github.com/dafny-lang/dafny/pull/1234)
```
See comment below for the formatting.
Besides the extra newline that is not necessarily ([have a look at here to see how release notes are currently formatted](https://github.com/dafny-lang/dafny/releases))I would also add that you would now be forcing users (and me) to do something that we don't want: Add a period before the parentheses.
So, in your example, and in your README, could you be specific about this point? Alternatively (and that would be my favorite method), you could detect the last character of the trimmed string, and if it's a period, you'd remove it.

Reply:
Done: I've removed the newline.  I find that it makes reading the release notes harder, but I won't fight about this.

The existing `RELEASE_NOTES` are inconsistent on whether to use a period,  so I've changed the code to add a period in all cases.  The version above with one period for the first sentence and no period for the second doesn't look right at all to me; do you like it better?

Reply:
>  I find that it makes reading the release notes harder

I won't argue since you implemented my change, thanks :-) but I'm just curious.
Can you please clarify, is it because you read the source of release notes instead of in a browser? Or is the browser rendering itself harder to read? To me, the first image here:
![image](https://user-images.githubusercontent.com/3601079/194348860-f0a4ec9f-29d0-43b2-9b08-ffcd1796ae58.png)
is much nicer than the alternative that you might find more appealing based on your script:
![image](https://user-images.githubusercontent.com/3601079/194349491-2954529a-567d-4d6c-9585-24548261c0ef.png)
I find that the second one breaks the flow, it feels like every link is important when the only thing I would like out of release notes is to read them in diagonal.

Regarding two lines: I think we should not even consider the possibility that a feat/fix line in a release notes should have two sentences. I know you added support for it, but it's not something we have done in the past and I don't see a reason why we should embrace those. So if we were to add paragraphs, I'd agree with you, we should end every sentence with a full stop, and make the link its own period-less sentence after that.
However, for a single phrase, with the comment in parentheses containing the link being part of the phrase, there should be no period between the phrase and the parentheses. If there was a period, it should come afterwards.

Now, one more principle to consider: We use bullet points to express each individual release notes, and many bullet points are not complete sentences but just fragments. In this case I think it's [conventional wisdom](https://www.instructionalsolutions.com/blog/bulleted-list-punctuation) to omit the period at the end of the sentence (after the parentheses containing the link).

Reply:
Good point.  I implemented your suggestion below: no auto punctuation fix and no newline before link, unless there's already multiple lines in the entry.

We can refine the rendering further, probably when we have more data (before the next release)